src/lib/operators/join_hash.cpp

```diff
@@ -82,6 +84,50 @@ void JoinHash::_on_cleanup() { _impl.reset(); }
 // currently using 32bit Murmur
 using Hash = uint32_t;
 
+// JoinHashTraits. Use left type for hashing by default
+template <typename L, typename R, class Enable = void>
+struct JoinHashTraits {
+  typedef L hash_type;
+  static constexpr bool needs_lexical_cast = false;
+};
+
+// Same types
+template <typename L, typename R>
+struct JoinHashTraits<L, R, typename std::enable_if_t<std::is_same<L, R>::value, void>> {
+  typedef L hash_type;
+  static constexpr bool needs_lexical_cast = false;
+};
+
+// If both are floating types, use the larger type to hash
+template <typename L, typename R>
+struct JoinHashTraits<
+    L, R,
+    typename std::enable_if_t<
+        std::is_floating_point<L>::value && std::is_floating_point<R>::value && sizeof(L) < sizeof(R), void>> {
+  typedef R hash_type;
+  static constexpr bool needs_lexical_cast = false;
+};
+
+// If both are integer types, use the larger type to hash
+template <typename L, typename R>
+struct JoinHashTraits<
+    L, R,
+    typename std::enable_if_t<
+        std::numeric_limits<L>::is_integer && std::numeric_limits<R>::is_integer && sizeof(L) < sizeof(R), void>> {
+  typedef R hash_type;
+  static constexpr bool needs_lexical_cast = false;
+};
+
+// Joining strings with numbers is not supported at the moment
+template <typename L, typename R>
+struct JoinHashTraits<L, R,
+                      typename std::enable_if_t<!std::is_same<L, R>::value && (std::is_same<R, std::string>::value ||
+                                                                               std::is_same<L, std::string>::value),
+                                                void>> {
+  typedef std::string hash_type;
+  static constexpr bool needs_lexical_cast = true;
+};
+
```

Comment:
This is cool! Here are some remarks:

- I’m not sure if the code leads to the desired behaviour at the moment. If we join an integral and a floating point type, we always take the left type. Is this what we want? 

- Why do strings require a lexical cast to string?

- `hash_type` is a type and should be written `HashType`

- I suggest removing everything that isn’t necessary or makes things longer than they need to be. For example:
  - Use `std::is_same_v`
  - Remove `typename` in front of `std::enable_if_t`
  - Remove template parameter `void` from `std::enable_if_t` (it’s the default)
  - the specialization for equal types is covered by the rest.



- Prefer `using` over `typedef`. The inversed order of arguments confuses me.

- Use `std::is_integral_v` instead of `std::numeric_limits<>`

- For picking the larger type I suggest using `std::conditional_t`
  ```c++
  template <typename L, typename R>
  struct JoinHashTraits<L, R, std::enable_if_t<std::is_integral_v<L> && std::is_integral_v<R>>> {
    using HashType = std::conditional_t<sizeof(L) < sizeof(R), R, L>;
  };
  ```

- For strings you first check if they are the same and then you check if either is of type `std::string`. Checking one of them is enough.

- I think we should think about future types, too. If someone added a new type, the default case would be chosen automatically without warning. I think it’s better to implement no default case at all so that who ever adds new types is forced to think about how the JoinHash should handle the new type.

Reply:
thank you for this fantastic feedback! I believe I was able to put every in. The traits have been moved to a separate file so that I can also test them thoroughly